### PR TITLE
Extend Craystack API allowing adaptive codecs, add Multiset codec, and extra example.

### PR DESCRIPTION
**Pull request by dsevero**
# Adaptive codecs
As mentioned in the title, this would allow adaptive codecs to be implemented in Craystack.

Signatures of push and pop are extended to allow for a `*context` variable.
```
    push: (ans_state, symbol, *context) -> (ans_state, *context)
    pop: (ans_state, *context) -> (ans_state, symbol, *context)
```
Note that, since context is passed via unpacking (i.e. *context), then it is essentially optional. Therefore, previous codecs are still compliant with this new signature. 

However, the return of `push` will be at least `(ans_state,)`, so function calls have been adapted accordingly. This means that instead of
```python
message = codec.push(message, symbol)
```
we now have
```python
message, = codec.push(message, symbol)
```

# Multiset codec
The multiset codec from https://github.com/facebookresearch/multiset-compression was ported to `craystack/codecs/multiset.py`

# Extra example
An extra example, showing how to use the multiset codec for BMNIST with BB-ANS, was added to `craystack/examples`.

### CODE REVIEW

File: craystack/codecs_test.py [hunk 1]
```diff
@@ -166,7 +166,7 @@ def test_categorical_new():
     rng = np.random.RandomState(2)
     precision = 4
     shape = (20, 3, 5)
-    weights = rng.random((np.prod(shape), 4))
+    weights = rng.random((np.prod(shape), 4)) + 1
```
Comment (dsevero) [inline craystack/codecs_test.py:169]: The lack of the `+ 1` was sometimes causing the test to fail, as some weights would be quantized to `0` due to low precision.

File: examples/binary_mnist_vae_multiset.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+## Decode
+decode_t0 = time.time()
+message = cs.unflatten(flat_message, obs_size + latent_size)
+
+message, multiset_decoded = vae_pop(message, multiset_size=num_batches)
+decode_t = time.time() - decode_t0
+
+print('All decoded in {:.2f}s.'.format(decode_t))
+
+assert cs.check_multiset_equality(multiset, multiset_decoded)
```
Comment (dsevero) [inline examples/binary_mnist_vae_multiset.py:111]: I removed the message equality check, as it is incompatible with empty pop.

File: examples/binary_mnist_vae.py [hunk 3]
```diff
@@ -85,4 +81,3 @@ def vae_view(head):
 print('All decoded in {:.2f}s.'.format(decode_t))
 
 np.testing.assert_equal(images, images_)
-np.testing.assert_equal(message, init_message)
```
Comment (dsevero) [inline examples/binary_mnist_vae.py:88]: I removed the message equality check, as it is incompatible with empty pop.

File: examples/binary_mnist_vae.py [hunk 4]
```diff
@@ -16,7 +17,7 @@
 
 num_images = 10000
 num_pixels = num_images * 784
-batch_size = 10
```
Comment (dsevero) [inline examples/binary_mnist_vae.py:19]: Changed to be comparable with the multiset codec. However, not really necessary.

File: README.md [hunk 5]
```diff
@@ -1,6 +1,6 @@
 __Informal disclaimer:__ Craystack is experimental software for prototyping and research in lossless compression. We will change (and probably break) parts of Craystack, and do not make any guarantees about API stability.
 
-# Craystack [![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.4572729.svg)](https://doi.org/10.5281/zenodo.4572729)
+# Craystack [![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.4572729.svg)](https://doi.org/10.5281/zenodo.4572729) [![craystack](https://github.com/dsevero/craystack/actions/workflows/tests.yml/badge.svg)](https://github.com/dsevero/craystack/actions/workflows/tests.yml)
```
Comment (j-towns) [inline README.md:3]: Presumably this link should point to the main branch rather than your fork?

File: craystack/codecs/multiset.py [hunk 6]
```diff
@@ -0,0 +1,205 @@
+"""
+Copyright (c) Facebook, Inc. and its affiliates.
+
+This source code is licensed under the MIT license found in the
+LICENSE file in the root directory of this source tree.
```
Comment (j-towns) [inline craystack/codecs/multiset.py:5]: You might need to append the contents of https://github.com/facebookresearch/multiset-compression/blob/master/LICENSE to Craystack's LICENSE.

**Comment by j-towns** [comment]
Very nice work. I'm actually wondering if we shouldn't go a bit further with the API change, and allow push and pop to be _any_ inverse pair, i.e. instead of
```
    push: (ans_state, symbol, *context) |-> (ans_state, *context)
    pop: (ans_state, *context) |-> (ans_state, symbol, *context)
```
we go for
```
    push: before |-> after
    pop: after |-> before
```
There's a trade-off here — instead of having to _unpack_ results, as in
```python
message, = codec.push(message, symbol)
```
the 'inverses' approach would require _packing_ inputs, as in
```python
message = codec.push((message, symbol))
```
because each push/pop must have exactly one argument.

An advantage of the 'inverses' approach is that the codec combinators (things like `serial` and `parallel`) might be simpler, and it might reveal some more general forms. For example, what we now call `repeat`, might be replaced by an 'invertible reduce':
```python
from functools import reduce

def ireduce(codec, length):
    # Assumes that
    #   codec.push: (a, b) |-> a
    #   codec.pop: a |-> (a, b)
    # and returns a codec for lists containing type b elements:
    #   push: (a, bs) |-> a
    #   pop: a |-> (a, bs)
    def push(a, bs):
        return reduce(codec.push, bs, a)

    def pop(a):
        bs = []
        for _ in range(length):
            a, b = codec.pop(a)
            bs.append(b)
            del b
        return bs
    return codec(push, pop)
```
This is a change we could make in a separate pr/series of prs.

**Comment by dsevero** [comment]
Maybe that could be a lower layer, and the codecs could be implemented on top? I'm worried that it might be a bit much for the regular user that just wants to do compression.

**Pull request merged**